### ISSUE TITLE
Add support for image data with NaNs

### ISSUE THREAD

**Issue by michaelaye** [scikit-image/scikit-image#1624]
Is there a general strategy on how to cope with it? Is it a goal to improve most tools to be able to cope sanely with image data that contains NaNs?

Currently found `disturbers':
- all Otsu tools
- random_walker segmentation

I could help with this, if wished, but maybe this came up before and it was decided to not have this as a goal for skimage?

**Comment by JDWarner** [comment]
@michaelaye, there has been some brief chatter about NaNs but no real impetus to make it happen. I think a large part of this is twofold:

* Unlike other fields of study (statistics, medicine), missing data is relatively uncommon in imaging by virtue of how images are typically acquired
* There is not, to my knowledge, broad agreement about how missing image data _should_ be handled - and it may well be so application-specific that attempting to generalize it would be ill-advised

Also, implementing NaN support would impose a significant computational burden. Many of our best algorithms need all the help they can get, from a performance standpoint.

In a general sense, many algorithms require or inherently expect a densely sampled array. You can't do traditional Fourier analysis easily with missing values, for example. Which means you must either figure out how to do that, fill the missing values (interpolation? median of nearest neighbors? chaos?) or extrapolate them, if they are not bounded. Danger lies in all directions; the steps along the way remain open fields of research.

Non-Fourier-based analyses also often require knowledge of a neighborhood. It's theoretically possible to modify these to be happy and adapt, using the rest of the neighborhood. But here be dragons too, e.g., zero division if if a lonely pixel has no or few neighbors. And of course, performance.


**I think supporting NaNs would be _fantastic_, for the record!** I'm not trying to rain on any parades... the points above can potentially be addressed. Just make sure you know what you're getting in to ;)

**Comment by michaelaye** [comment]
Thanks for your feedback. How about the Otsu stuff? As they are just 1-D calculations, should that not be okay to just filter out NaNs automatically for that?

**Comment by JDWarner** [comment]
@michaelaye Indeed, those would be an ideal place to begin! Simply omitting NaNs seems perfectly appropriate for non-adaptive thresholding (adaptive may be a little trickier).

**Comment by ananana** [comment]
I'd like to work on this at the sprints

**Comment by local-minimum** [comment]
@ananana I don't know when the sprints occur or if you've started working on this issue, else I could add support for `numpy.ma.masked_array` input to the `filters.thresholding.otsu_threshold`. @michaelaye Are there other Otsu-functions I should look at if I'm doing this?

In general, I suggest not implicitly filtering out `nan` out of image data but require masking, to be sure user code errors are still visible/cause exceptions/result in nan and not silently handled by filtering any non-finite data. It should also make tests quite easy to construct.

I'm new to contributing to others' gits so I'm a not sure, but if I'm doing this, I suppose I should just make my own fork and reference this issue in the pull-request?

**Comment by soupault** [comment]
@local-minimum we've a [nice guide](http://scikit-image.org/docs/dev/contribute.html) to facilitate new contributions.

Concerning masked arrays:
I think we need to work out some kind of specification on what should be the _correct_ use of masked array from both user and our points of view. I think, in `numpy` `masked` stands for `missing`.
At the same time, from my understanding, in image processing (as Josh said) missing data appears in several very specific cases:
- bad-pixels inpainting, sensors' butting gap correction;
- demosaic;
- segmented objects' manipulation;
- aftertouch, object removal.

Notice, that the meaning of mask is quite different across algorithms.
Saying this, there is not much sense to me to carry out masked array along the full processing pipeline. I think, it is much clearer to pass `mask` as an argument explicitly to those several functions, which may benefit from it (like, for example, [inpainting](https://github.com/scikit-image/scikit-image/blob/master/skimage/restoration/inpaint.py#L75)).

**Comment by local-minimum** [comment]
I'll read the faq on contributions.

As for the question at hand, I see the argument for supplying `mask` rather than actually requiring a `numpy.ma.masked_array`.

However passing a `masked_array` to `threshold_utso` at the moment will work if the mask is all `False` and of int type but if there is any `True` or of float type it will cause a `TypeError`, at least in Python 2.7. This is a rather unexpected behaviour.

Could one maybe do the following:

    if mask is not None:
        image = image[mask]
    elif isinstance(image, numpy.ma.masked_array):
        image = image[mask == numpy.False_]

**Comment by jni** [comment]
@soupault 

> Answering to the comment above by @local-minimum , I'd still prefer to keep True for invalid, False for valid);

This is counter to how much of the SciPy ecosystem works. (Not to mention our existing behaviour when we have a `mask=` keyword argument.) So I am :-1: on this proposal and suggest that `True` means 'valid'. One example where this is useful is where using a mask both for indexing (e.g. to do manual processing) and for a masked operation. Or for labeling. It is nice not to have to invert the mask here.

**Comment by soupault** [comment]
>This is counter to how much of the SciPy ecosystem works.

@jni could you, please, provide any pointers to support this statement? When developing `inpaint`, I was taking inspiration from this guide - https://docs.scipy.org/doc/numpy-1.13.0/reference/maskedarray.generic.html#what-is-a-masked-array : 
```
A mask is either nomask, indicating that no value of the associated array is invalid,
or an array of booleans that determines for each element of the associated array
whether the value is valid or not. When an element of the mask is False, the 
corresponding element of the associated array is valid and is said to be unmasked. 
When an element of the mask is True, the corresponding element of the associated 
array is said to be masked (invalid).
```

**Comment by jni** [comment]
@soupault 

> I was taking inspiration from this guide

Yes, but NumPy masked arrays are poorly supported and face an uncertain future. Boolean indexing does not. In addition, from [`scipy.ndimage.binary_dilation`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.binary_dilation.html#scipy.ndimage.binary_dilation):

> mask : array_like, optional
>     If a mask is given, only those elements with a True value at the corresponding mask element are modified at each iteration.

This is true of many other morphology functions in `ndimage`. It's also true of our own functions, such as `watershed` and many of our filters (e.g. `filters.sobel`). 

It's not clear from the documentation, but [`sklearn.feature_extraction.image.img_to_graph`](http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.image.img_to_graph.html) uses `mask` in this way also (True is valid).

As a counterpoint, astropy [uses a subclass](http://docs.astropy.org/en/stable/table/masking.html) of `MaskedArray` and so uses the other convention.

But, here's another part of the astropy project, ["Spectral Cube"](https://spectral-cube.readthedocs.io/en/latest/masking.html), that uses my preferred convention:

> In addition to supporting the representation of data and associated WCS, it is also possible to attach a boolean mask to the SpectralCube class. Masks can take various forms, but one of the more common ones is a cube with the same dimensions as the data, and that contains e.g. the boolean value True where data should be used, and the value False when the data should be ignored

Overall, the most compelling argument for me is that True=valid is the convention most convenient for working with standard NumPy arrays, since mostly you want to do things to pixels inside the mask, not outside of it, so you can use the mask directly instead of having to invert it all the time.

**Comment by lagru** [comment]
I'm removing this from the 1.0 milestone. From skimming this it does not seem like there are concrete actionable items yet. I gather that we mostly adere to using `True` == valid in most places of our library now (at least that's my general experience). Otherwise https://github.com/scikit-image/scikit-image/issues/1624#issuecomment-419654966 seems quite close to the current consensus.
